gunrock/graph/dynamic_graph/dynamic_graph_parallel_iterator.cuh

```diff
@@ -0,0 +1,162 @@
+// ----------------------------------------------------------------------------
+// Gunrock -- Fast and Efficient GPU Graph Library
+// ----------------------------------------------------------------------------
+// This source code is distributed under the terms of LICENSE.TXT
+// in the root directory of this source distribution.
+// ----------------------------------------------------------------------------
+
+/**
+ * @file
+ * dynamic_graph_parallel_iterator.cuh
+ *
+ * @brief Dynamic Graph Data Structure iterator
+ */
+#pragma once
+
+#include <gunrock/util/array_utils.cuh>
+#include <gunrock/graph/csr.cuh>
+#include <gunrock/graph/graph_base.cuh>
+
+#include <gunrock/graph/dynamic_graph/dynamic_graph_base.cuh>
+#include <gunrock/graph/dynamic_graph/dynamic_graph_unweighted.cuh>
+#include <gunrock/graph/dynamic_graph/dynamic_graph_weighted.cuh>
+
+namespace gunrock {
+namespace graph {
+
+template <typename VertexT, typename ValueT, bool IsMap>
+struct nodeReader {
+  __device__ static void readNode(const uint32_t& key, VertexT& eId,
+                                  ValueT& eVal, const int& laneId) {
+    eVal = __shfl_xor_sync(0xFFFFFFFF, key, 1);
+    bool key_lane = !(laneId % 2) && (laneId < 30);
+    eId = key_lane ? key : EMPTY_KEY;
+    eVal = key_lane ? eVal : EMPTY_KEY;
+    eId = (key != EMPTY_KEY) ? eId
+                             : util::PreDefinedValues<VertexT>::InvalidValue;
+    eVal = (key != EMPTY_KEY) ? eVal
+                              : util::PreDefinedValues<ValueT>::InvalidValue;
+  }
+};
+template <typename VertexT, typename ValueT>
+struct nodeReader<VertexT, ValueT, false> {
+  __device__ static void readNode(const uint32_t& key, VertexT& eId,
+                                  ValueT& eVal, const int& laneId) {
+    bool key_lane = laneId < 30;
+    eId = key_lane ? key : EMPTY_KEY;
+    bool is_valid = (key != EMPTY_KEY);
+    eId = (key != EMPTY_KEY) ? eId
+                             : util::PreDefinedValues<VertexT>::InvalidValue;
+    eVal = util::PreDefinedValues<ValueT>::InvalidValue;
+  }
+};
+
+/**
+ * @brief ParallelIterator iterator for dynamic graph.
+ *
+ * @tparam VertexT Vertex identifier type.
+ * @tparam SizeT Graph size type.
+ * @tparam ValueT Associated value type.
+ * @tparam GraphFlag graph flag
+ */
+template <typename VertexT, typename SizeT, typename ValueT, GraphFlag FLAG>
+struct ParallelIterator<VertexT, SizeT, ValueT, FLAG, HAS_DYN> {
```

Comment:
Similar to Csr, does this file need to exist? Does anything get simpler if the iterator lived inside the thing it iterates over?

Reply:
Same as above. I personally prefer separating graph files into three parts: 1) the basic class code, 2) the iterator on top of that, 3) the algorithms on top of 1, 2. But I'm open to ideas here as well.

Reply:
Same as above -- let's roll with what you have now.